assets/js/googlesitekit/modules/datastore/sharing-settings.js

```diff
@@ -221,15 +247,175 @@ const baseReducer = ( state, { type, payload } ) => {
+	 */
+	getShareableRoles( state ) {
+		const { shareableRoles } = state;
+		return shareableRoles;
+	},
+
+	/**
+	 * Gets the dashboard sharing management for the given module.
+	 *
+	 * Returns the module's sharing management string.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state      Data store's state.
+	 * @param {string} moduleSlug Module slug.
+	 * @return {(string|null|undefined)} The module's sharing management string, null if there is none,
+	 *                                   undefined if not loaded yet.
+	 */
+	getSharingManagement: createRegistrySelector(
+		( select ) => ( state, moduleSlug ) => {
+			invariant( moduleSlug, 'moduleSlug is required.' );
+			const sharingSettings = select( CORE_MODULES ).getSharingSettings();
+
+			if ( sharingSettings === undefined ) {
+				return undefined;
+			}
+			return sharingSettings[ moduleSlug ]?.management || null;
+		}
+	),
+
+	/**
+	 * Gets the shared roles for the given module.
+	 *
+	 * Returns the module's shared roles list.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state      Data store's state.
+	 * @param {string} moduleSlug Module slug.
+	 * @return {(Array|null|undefined)} The module's shared roles array, null if there is none,
+	 *                                   undefined if not loaded yet.
+	 */
+	getSharedRoles: createRegistrySelector(
+		( select ) => ( state, moduleSlug ) => {
+			invariant( moduleSlug, 'moduleSlug is required.' );
+			const sharingSettings = select( CORE_MODULES ).getSharingSettings();
+
+			if ( sharingSettings === undefined ) {
+				return undefined;
+			}
+			return sharingSettings[ moduleSlug ]?.sharedRoles || null;
+		}
+	),
+
+	/**
+	 * Indicates whether the current sharing settings have changed from what is saved.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object}     state Data store's state.
+	 * @param {Array|null} keys  Sharing Settings keys to check; if not provided, all sharing settings are checked.
+	 * @return {boolean} True if the sharing settings have changed, false otherwise.
+	 */
+	haveSharingSettingsChanged( state, keys = null ) {
+		const { sharingSettings, savedSharingSettings } = state;
+
+		if ( keys ) {
+			return ! isEqual(
+				pick( sharingSettings, keys ),
+				pick( savedSharingSettings, keys )
+			);
+		}
+
+		return ! isEqual( sharingSettings, savedSharingSettings );
+	},
+
+	/**
+	 * Checks whether we can submit sharing settings changes for a module.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state      Data store's state.
+	 * @param {string} moduleSlug Module slug.
+	 * @return {boolean} Whether the module supports submitting sharing settings changes.
+	 */
+	canSubmitSharingChanges: createRegistrySelector(
+		( select ) => ( state, moduleSlug ) => {
+			invariant( moduleSlug, 'moduleSlug is required.' );
+			const storeName = select( CORE_MODULES ).getModuleStoreName(
+				moduleSlug
+			);
+			return !! select( storeName )?.canSubmitChanges?.();
+		}
+	),
```

Comment:
Looks like there is an inaccuracy in IB for this selector. This selector shouldn't call `canSubmitChanges` selector for a module, it doesn't make sense. What we need to do here is to check whether the shared settings have changed and to use the approach similar to one that we use for `canSubmitChanges` where we have "safe" and "dangerous" versions of the `canSubmitChanges` selector.

Learn how we use the `createValidationSelector` function for the `createSubmitChangesStore` factory and the `makeDefaultCanSubmitChanges` function for default implementation. The main difference is that for this PR the `canSubmitSharingChanges` implementation should be module agnostic.

Also, pay attention that you will need to add a selector similar to `isDoingSubmitChanges` to check whether we are submitting sharing changes or not. Once again, you can see it's implementation in the `createSubmitChangesStore` factory.

cc @aaemnnosttv in case I miss something.

Reply:
> Looks like there is an inaccuracy in IB for this selector. This selector shouldn't call `canSubmitChanges` selector for a module, it doesn't make sense. What we need to do here is to check whether the shared settings have changed and to use the approach similar to one that we use for `canSubmitChanges` where we have "safe" and "dangerous" versions of the `canSubmitChanges` selector.

@eugene-manuilov correct, the ACs and IB only mention to use `canSubmitChanges` as an example, it isn't relevant to call this here as you said.

> Also, pay attention that you will need to add a selector similar to `isDoingSubmitChanges` to check whether we are submitting sharing changes or not.

I'm not sure if we need this for sharing settings yet. This is more relevant for module settings because there may be multiple requests needed in order to submit all the changes. In the case of sharing settings, we should be able to get by with just the `isFetching*` selector generated by the `saveSharingSettings` fetch store. We can always wrap it later if needed with the same kind of abstraction if needed.

Reply:
Chatting with @hussain-t about this, we decided to keep the `isDoingSubmitChanges` selector as the wrapping actions are already there in the save action, it just needs the reducer cases and selector as a simple getter. Relying on the `isFetching` action could be problematic as owner IDs will change after the request is complete if owners change.